### PR TITLE
Static functions, index by number

### PR DESCRIPTION
**Pull request by HayrapetyanMurad**
Add possibility to add static functions to the class (which enables adding metamethods to the class)
Add possibility to set index by number functions

- - -
This change is [Reviewable](https://reviewable.io/reviews/PicsArt/luabind/6)

### CODE REVIEW

File: include/luabind/wrapper.hpp [hunk 1]
```diff
@@ -162,6 +162,40 @@ struct function_wrapper<R (*)(Args...), func> {
     }
 };
 
+template <typename F, F f>
+struct static_function_wrapper;
+
+template <typename R, typename... Args, R (*func)(Args...)>
+struct static_function_wrapper<R (*)(Args...), func> {
```
Comment (davits) [inline include/luabind/wrapper.hpp]: Why this class was added? the function_wrapper covers this case fully.

File: include/luabind/bind.hpp [hunk 2]
```diff
@@ -84,6 +84,26 @@ class class_ {
         return *this;
     }
 
+    template <auto func>
+    class_& static_function(const char* name) {
```
Comment (davits) [inline include/luabind/bind.hpp]: Lets rename this to class_function ... It is more correct terminology generic language wise.

File: include/luabind/bind.hpp [hunk 3]
```diff
@@ -147,15 +178,21 @@ class class_ {
...
+        if (lua_type(L, 2) == LUA_TNUMBER && info->index_by_number) {
+            return info->index_by_number(L);
         }
-        auto f_it = info->functions.find(std::string_view(key));
-        if (f_it != info->functions.end()) {
-            lua_pushcfunction(L, f_it->second);
-            return 1;
+
+        if (lua_type(L, 2) == LUA_TSTRING) {
+            const char* key = lua_tostring(L, 2);
```
Comment (davits) [inline include/luabind/bind.hpp]: Use `value_mirror<std::string_view>::from_lua(L, 2);` it is faster.

File: include/luabind/bind.hpp [hunk 4]
```diff
@@ -183,16 +220,23 @@ class class_ {
...
-                return 1;
-            } else {
-                luaL_error(L, "property named '%s' is read only", key);
+
+        if (lua_type(L, 2) == LUA_TNUMBER && info->newindex_by_number) {
+            return info->newindex_by_number(L);
+        }
+
+        if (lua_type(L, 2) == LUA_TSTRING) {
+            const char* key = lua_tostring(L, 2);
```
Comment (davits) [inline include/luabind/bind.hpp]: Here too.

**Review by davits — CHANGES_REQUESTED** [review]

File: include/luabind/bind.hpp [hunk 5]
```diff
@@ -124,6 +144,17 @@ class class_ {
         return *this;
     }
 
+    class_& index_by_number(lua_CFunction getter_function) {
```
Comment (davits) [inline include/luabind/bind.hpp]: I am thinking about more user friendly name...
`.array(get, set)` ???

File: include/luabind/user_data.hpp [hunk 6]
```diff
@@ -59,6 +59,10 @@ class user_data {
         lua_rawset(L, table_idx);
     }
 
+    static void get_custom_table(lua_State* L, int idx) {
```
Comment (davits) [inline include/luabind/user_data.hpp:62]: Please also add `set_custom_table()` function, which pops value from stack and sets it as a custom table for value at idx.

File: include/luabind/bind.hpp (hunk 2 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp]: done

File: include/luabind/bind.hpp (hunk 3 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp]: done

File: include/luabind/bind.hpp (hunk 4 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp]: done

File: include/luabind/bind.hpp (hunk 5 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp]: maybe array_access(get,set)

File: include/luabind/user_data.hpp (hunk 6 above)
Comment (HayrapetyanMurad) [reply include/luabind/user_data.hpp:62]: done

File: include/luabind/wrapper.hpp (hunk 1 above)
Comment (HayrapetyanMurad) [reply include/luabind/wrapper.hpp]: function_wrapper has a check that the argument count should be sizeof...(Args) but for class functions it should be sizeof...(Args) + 1, as user will call this functions as Class:func(args)

File: include/luabind/bind.hpp [hunk 7]
```diff
@@ -84,6 +84,26 @@ class class_ {
         return *this;
     }
 
+    template <auto func>
+    class_& class_function(const char* name) {
+        _info->get_metatable(_L);
+        lua_pushstring(_L, name);
+        lua_CFunction cwrapper = &class_function_wrapper<decltype(func), func>::invoke;
```
Comment (davits) [inline include/luabind/bind.hpp:89]: just call the function below with this argument.

File: include/luabind/bind.hpp (hunk 5 above)
Comment (davits) [reply include/luabind/bind.hpp]: Sounds good...

File: include/luabind/bind.hpp [hunk 8]
```diff
@@ -147,15 +178,21 @@ class class_ {
...
+            auto p_it = info->properties.find(key);
+            if (p_it != info->properties.end()) {
+                return p_it->second.getter(L);
+            }
+            auto f_it = info->functions.find(key);
+            if (f_it != info->functions.end()) {
+                lua_pushcfunction(L, f_it->second);
+                return 1;
+            }
         }
```
Comment (davits) [inline include/luabind/bind.hpp:172]: Please fail if the key is not a number nor a string.

File: include/luabind/type_storage.hpp [hunk 9]
```diff
@@ -36,6 +36,8 @@ struct type_info {
     lua_CFunction new_index;
     lua_CFunction subscript_read;
```
Comment (davits) [inline include/luabind/type_storage.hpp]: subscript_read/write are the leftovers from my attempt at implementing array access, please remove those.

Comment (HayrapetyanMurad) [reply include/luabind/type_storage.hpp]: Done

File: include/luabind/bind.hpp (hunk 8 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp:172]: Done.

File: include/luabind/bind.hpp (hunk 7 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp:89]: done

File: include/luabind/bind.hpp (hunk 5 above)
Comment (HayrapetyanMurad) [reply include/luabind/bind.hpp]: done

**Review by davits — APPROVED** [review]

**Pull request merged**